Cannot use a 'SHARED' Kafka data connection to create a mapping [HZ-2392]

<!--
Thanks for reporting your issue. Please share with us the following information, to help us resolve your issue quickly and efficiently.
-->
  
**Describe the bug**
After https://github.com/hazelcast/hazelcast/pull/24333, it seems it's no longer possible to use shared `SHARED` data connection for Kafka. If the user tries to provide options via `CREATE DATA CONNECTION...` only, they will get `com.hazelcast.sql.HazelcastSqlException: Missing 'valueFormat' option` when attempting to create a mapping from it. If they provide the `keyFormat` and `valueFormat` options when creating the mapping, the user will get the following error when trying to use this mapping in a query:

```
com.hazelcast.core.HazelcastException: Shared Kafka producer can be created only with data connection options
com.hazelcast.jet.JetException: Exception in ProcessorTasklet{my_job/Kafka[public.my_topic]#0}: com.hazelcast.core.HazelcastException: Shared Kafka producer can be created only with data connection options
        at com.hazelcast.jet.impl.execution.TaskletExecutionService.handleTaskletExecutionError(TaskletExecutionService.java:286)
        at com.hazelcast.jet.impl.execution.TaskletExecutionService.access$600(TaskletExecutionService.java:80)
        at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:325)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
Caused by: com.hazelcast.core.HazelcastException: Shared Kafka producer can be created only with data connection options
```

**Expected behavior**
The mapping can be creating w/o providing the options (i.e. options are provided when creating the shared Kafka DC)

**To Reproduce**

The following unit test can be pasted into `KafkaDataConnectionIntegrationTest` to demonstrate the issue:

```
    @Test
    public void test_canCreateShared() {
        String dcName = randomName();
        String name = createRandomTopic(PARTITION_COUNT);
        createSqlKafkaDataConnection(dcName, true);
        createKafkaMappingUsingDataConnection(name, dcName, "");

        try (SqlResult r = sqlService.execute("INSERT INTO " + name + " VALUES (0, 'value-0')")) {
            assertEquals(0, r.updateCount());
        }

        assertTipOfStream("SELECT * FROM " + name, singletonList(new Row(0, "value-0")));
    }
```

--- a/extensions/kafka/src/main/java/com/hazelcast/jet/kafka/KafkaDataConnection.java
+++ b/extensions/kafka/src/main/java/com/hazelcast/jet/kafka/KafkaDataConnection.java
@@ -193,7 +193,10 @@ public class KafkaDataConnection extends DataConnectionBase {
 
         boolean inputPropsAreSubsetOfConfigProps = inputPropsAreSubsetOfConfigProps(properties);
         if (getConfig().isShared() && !properties.isEmpty() && !inputPropsAreSubsetOfConfigProps) {
-            throw new HazelcastException("Shared Kafka producer can be created only with data connection options");
+            throw new HazelcastException("For shared Kafka producer, please provide all serialization options" +
+                    "at the DATA CONNECTION level (i.e. 'key.serializer')." +
+                    " Only 'keyFormat' and 'valueFormat' are required at the mapping level," +
+                    " however these options are ignored currently.");
         }
 
         Properties props = configProperties;

--- a/extensions/kafka/src/main/java/com/hazelcast/jet/kafka/KafkaDataConnection.java
+++ b/extensions/kafka/src/main/java/com/hazelcast/jet/kafka/KafkaDataConnection.java
@@ -189,15 +189,23 @@ public class KafkaDataConnection extends DataConnectionBase {
     public <K, V> KafkaProducer<K, V> getProducer(
             @Nullable String transactionalId,
             @Nonnull Properties properties) {
-        if (getConfig().isShared() && !properties.isEmpty()) {
+        Properties configProperties = getConfig().getProperties();
+
+        boolean inputPropsAreSubsetOfConfigProps = inputPropsAreSubsetOfConfigProps(properties);
+        if (getConfig().isShared() && !properties.isEmpty() && !inputPropsAreSubsetOfConfigProps) {
             throw new HazelcastException("Shared Kafka producer can be created only with data connection options");
-        } else {
-            Properties props = Util.mergeProps(getConfig().getProperties(), properties);
-            if (transactionalId != null) {
-                props.put("transactional.id", transactionalId);
-            }
-            return new KafkaProducer<>(props);
         }
+
+        Properties props = configProperties;
+        // If argument properties map is a subset of data connection properties map, just skip that merge.
+        if (!getConfig().isShared() || inputPropsAreSubsetOfConfigProps) {
+            props = Util.mergeProps(configProperties, properties);
+        }
+
+        if (transactionalId != null) {
+            props.put("transactional.id", transactionalId);
+        }
+        return new KafkaProducer<>(props);
     }
 
     @Override
@@ -206,5 +214,15 @@ public class KafkaDataConnection extends DataConnectionBase {
             producerSupplier.remembered().doClose();
             producerSupplier = null;
         }
+    }
+
+    private boolean inputPropsAreSubsetOfConfigProps(Properties inputProps) {
+        Properties configProperties = getConfig().getProperties();
+        for (Object key : inputProps.keySet()) {
+            if (!configProperties.containsKey(key) || !configProperties.get(key).equals(inputProps.get(key))) {
+                return false;
+            }
+        }
+        return true;
     }
 }

--- a/extensions/kafka/src/test/java/com/hazelcast/jet/kafka/KafkaDataConnectionTest.java
+++ b/extensions/kafka/src/test/java/com/hazelcast/jet/kafka/KafkaDataConnectionTest.java
@@ -196,7 +196,7 @@ public class KafkaDataConnectionTest {
 
         assertThatThrownBy(() -> kafkaDataConnection.getProducer(null, properties))
                 .isInstanceOf(HazelcastException.class)
-                .hasMessageContaining("Shared Kafka producer can be created only with data connection options");
+                .hasMessageContaining("For shared Kafka producer, please provide all serialization options");
 
         kafkaDataConnection.release();
     }

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/KafkaDataConnectionIntegrationTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/KafkaDataConnectionIntegrationTest.java
@@ -17,6 +17,7 @@
 package com.hazelcast.jet.sql.impl.connector.kafka;
 
 import com.hazelcast.core.HazelcastException;
+import com.hazelcast.sql.SqlResult;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
@@ -24,7 +25,11 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_FORMAT;
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_FORMAT;
 import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
 
 @RunWith(HazelcastSerialClassRunner.class)
@@ -34,7 +39,7 @@ public class KafkaDataConnectionIntegrationTest extends KafkaSqlTestSupport {
     private static final int PARTITION_COUNT = 1;
 
     @Test
-    public void test_nonShared() {
+    public void when_createNonSharedProducer_then_success() {
         String dlName = randomName();
         String name1 = createRandomTopic(PARTITION_COUNT);
         String name2 = createRandomTopic(PARTITION_COUNT);
@@ -86,19 +91,51 @@ public class KafkaDataConnectionIntegrationTest extends KafkaSqlTestSupport {
     }
 
     @Test
-    public void test_shared() {
+    public void when_createSharedProducer_then_success() {
+        String dcName = randomName();
+        String name = createRandomTopic(PARTITION_COUNT);
+        createSqlKafkaDataConnection(dcName, true);
+        createKafkaMappingUsingDataConnection(name, dcName, constructMappingOptions("int", "varchar"));
+
+        try (SqlResult r = sqlService.execute("INSERT INTO " + name + " VALUES (0, 'value-0')")) {
+            assertThat(r.updateCount()).isZero();
+        }
+
+        assertTipOfStream("SELECT * FROM " + name, singletonList(new Row(0, "value-0")));
+    }
+
+    @Test
+    public void when_createSharedProducerWithOverriddenProperty_then_success() {
         String dlName = randomName();
-        String name1 = createRandomTopic(PARTITION_COUNT);
-        String name2 = createRandomTopic(PARTITION_COUNT);
+        String name = createRandomTopic(PARTITION_COUNT);
         createSqlKafkaDataConnection(dlName, true);
-        createKafkaMappingUsingDataConnection(name1, dlName, constructMappingOptions("int", "varchar"));
-        createKafkaMappingUsingDataConnection(name2, dlName, constructMappingOptions("varchar", "int"));
+        // not created
+        createKafkaMappingUsingDataConnection(name, dlName,
+                "OPTIONS ('" + OPTION_KEY_FORMAT + "'='int', '" + OPTION_VALUE_FORMAT + "'='varchar', "
+                        + "'auto.offset.reset' = 'latest')"); // changed option
 
         // TODO: move this error to mapping creation level.
         assertThatThrownBy(() -> sqlService.execute(
-                        "INSERT INTO " + name1 + " VALUES" + "(0, 'value-0')"))
+                "INSERT INTO " + name + " VALUES" + "(0, 'value-0')"))
                 .hasRootCauseInstanceOf(HazelcastException.class)
                 .hasMessageContaining("Shared Kafka producer can be created only with data connection options");
 
+    }
+
+    /**
+     * <a href="https://github.com/hazelcast/hazelcast/issues/24283">Issue 24283</a>
+     */
+    @Test
+    public void when_creatingDataConnectionAndMapping_then_serdeDeterminesAutomatically() {
+        String dlName = randomName();
+        String name = createRandomTopic(PARTITION_COUNT);
+        createSqlKafkaDataConnection(dlName, false); // connection is not shared
+        createKafkaMappingUsingDataConnection(name, dlName, constructMappingOptions("int", "varchar"));
+
+        try (SqlResult r = sqlService.execute("INSERT INTO " + name + " VALUES (0, 'value-0')")) {
+            assertThat(r.updateCount()).isZero();
+        }
+
+        assertTipOfStream("SELECT * FROM " + name, singletonList(new Row(0, "value-0")));
     }
 }

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/KafkaDataConnectionIntegrationTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/KafkaDataConnectionIntegrationTest.java
@@ -118,7 +118,7 @@ public class KafkaDataConnectionIntegrationTest extends KafkaSqlTestSupport {
         assertThatThrownBy(() -> sqlService.execute(
                 "INSERT INTO " + name + " VALUES" + "(0, 'value-0')"))
                 .hasRootCauseInstanceOf(HazelcastException.class)
-                .hasMessageContaining("Shared Kafka producer can be created only with data connection options");
+                .hasMessageContaining("For shared Kafka producer, please provide all serialization options");
 
     }
 

